File: 6/6.2/HW 6 - 2/list.cpp
#include "list.h"
#include <iostream>
using namespace std;

struct ListElement
{
	int value;
	ListElement* next;
};

struct List
{
	ListElement* head;
};

List* createList()
{
	List* newList = new List;
	newList->head = nullptr;
	return newList;
}

void addElement(int x, List* list)
{
	ListElement* newValue = new ListElement;
	newValue->value = x;
	ListElement* temp = list->head;
	ListElement* futurePrevious = nullptr;
	while (true)
	{
		if (list->head == nullptr)
		{
			list->head = newValue;
			newValue->next = nullptr;
			break;
		}
		else if (temp->value > x && temp == list->head)
		{
			newValue->next = list->head;
			list->head = newValue;
			break;
		}
		else if (temp->value > x)
		{
			futurePrevious->next = newValue;
			newValue->next = temp;
			break;
		}
		else if (temp->next == nullptr)
		{
			newValue->next = nullptr;
			temp->next = newValue;
			break;
		}
		futurePrevious = temp;
		temp = temp->next;
	}
}

void deleteElement(int x, List* list)
{
	ListElement* temp = list->head;
	ListElement* previous = nullptr;
	while (temp != nullptr)
	{
		if (temp->value == x)
		{
			ListElement* temp2 = temp;
			if (temp == list->head)
			{
				list->head = list->head->next;
				temp = list->head;
			}
			else if (temp->next == nullptr)
			{
				previous->next = nullptr;
				temp = nullptr;
			}
			else
			{
				previous->next = temp->next;
				temp = temp->next;
			}
			delete temp2;
		}
		else
		{
			previous = temp;
			temp = temp->next;
		}
	}
}

void printList(List* list)
{
	ListElement* temp = list->head;
	while (temp != nullptr)
	{
		cout << temp->value << " ";
		temp = temp->next;
	}
	cout << endl;
}

File: hw2_2/hw2_2/List.cs
namespace List
{
    /// <summary>
    /// List.
    /// </summary>
    class List: IList
    {
        private int listSize = 0;
        private ListElement head = null;

        private class ListElement
        {
            public int Value { get; set; }

            public ListElement Next { get; set; }

            public ListElement()
            {
                Value = 0;
                Next = null;
            }

            public ListElement(int value, ListElement next)
            {
                Value = value;
                Next = next;
            }
        }

        public void AddElement(int element)
        {
            if (this.head == null)
            {
                this.head = new ListElement(element, null);
            }
            else if (this.head.Next == null)
            {
                this.head.Next = new ListElement(element, null);
            }
            else
            {
                ListElement temp = this.head;
                while (temp.Next != null)
                {
                    temp = temp.Next;
                }
                temp.Next = new ListElement(element, null);
            }
            ++this.listSize;
        }

        public bool IsEmpty() =>  this.head == null;

        public bool DeleteElement(int element)
        {
            if (this.IsEmpty())
            {
                return false;
            }

            ListElement temp = this.head;
            if (temp.Value == element)
            {
                this.head = this.head.Next;
                return true;
            }
            if (temp.Next != null)
            {
                if (temp.Next.Value == element)
                {
                    this.head.Next = this.head.Next.Next;
                    return true;
                }
            }
            else
            {
                return false;
            }

            while (temp.Next != null)
            {
                if (temp.Next.Value == element)
                {
                    temp.Next = temp.Next.Next;
                    return true;
                }
                temp = temp.Next;
            }
            return false;
        }
    }
}